File: paack-java-sdk/src/main/java/eu/paack/sdk/api/PodApi.java
package eu.paack.sdk.api;

import com.fasterxml.jackson.core.type.TypeReference;
import eu.paack.sdk.PaackConstants;
import eu.paack.sdk.api.model.response.Error;
import eu.paack.sdk.api.model.response.PaackResponse;
import eu.paack.sdk.api.model.response.ProofOfDeliveryResponse;
import eu.paack.sdk.exceptions.ApiException;
import lombok.NoArgsConstructor;
import lombok.experimental.SuperBuilder;
import lombok.extern.slf4j.Slf4j;
import org.apache.http.message.BasicNameValuePair;

import java.util.Collections;

/**
 * Proof of Delivery API allows you to easily validate that your orders have been delivered.
 * It works by extracting additional delivery verification data that is sent from the Paack driver application to Paack's database.
 *
 * More details can be found in the Paack API documentation: https://paack.readme.io/reference/orders-proof-of-delivery
 */
@Slf4j
@SuperBuilder()
@NoArgsConstructor
public class PodApi extends PaackApi {

    /**
     * Retrieves the delivery verifications for the retailer of the order external ID specified in the path parameter.
     * @param externalId
     * @return
     */
    public PaackResponse<ProofOfDeliveryResponse, Error> deliveryVerification(String externalId) {
        if (externalId == null || externalId.length() == 0) {
            return errorMessage("ExternalId cannot be empty", "externalId", "001");
        }
        try {
            BasicNameValuePair params = new BasicNameValuePair(PaackConstants.PARAM_EXTERNAL_ID, externalId);
            PaackResponse<ProofOfDeliveryResponse, Error> response = apiClient.invokeAPI(PaackEndpoint.pod,
                    "GET",
                    Collections.singletonList(params),
                    null,
                    null,
                    new TypeReference<ProofOfDeliveryResponse>() {
                    });
            log.info(response.toString());
            return response;
        } catch (ApiException e) {
            log.error("Delivery verification failed", e);
            return errorMessage("PodApi.deliveryVerification", "Delivery verification failed");
        }
    }
}

File: paack-go-sdk/rest/pod_api.go
package rest

import (
	"encoding/json"
	"fmt"
	"github.com/PaackEng/paack-go-sdk/api_client"
	"github.com/PaackEng/paack-go-sdk/logger"
	"github.com/PaackEng/paack-go-sdk/mixin"
	"github.com/PaackEng/paack-go-sdk/types/responses"
	"io"
	"strconv"
	"strings"
)

/*
PodApi - Proof of Delivery API allows you to easily validate that your orders have been delivered.
It works by extracting additional delivery verification data that is sent
from the Paack driver application to Paack's database.

More details can be found in the Paack API documentation: https://paack.readme.io/reference/orders-proof-of-delivery
*/
type PodApi struct {
	BaseApi
}

func NewPodApi(audience, domain string, path *mixin.Resources, client api_client.ApiClient, logger logger.Logger) *PodApi {

	return &PodApi{
		*NewBaseApi(client, domain, audience, path, logger),
	}
}

/*
DeliveryVerifications - Retrieves the delivery verifications for the retailer of the ID specified in the path parameter.
*/
func (t *PodApi) DeliveryVerifications(orderId string) (successResponse *responses.DeliveryVerificationsResponse, errorResponse *responses.ErrorResponse) {
	t.logger.Info("Calling PodApi.DeliveryVerifications method")

	// validate orderId
	if len(orderId) == 0 {
		errorResponse = responses.NewError(fmt.Sprintf("orderId cannot be empty"), "001", "orderId", t.logger)
		return
	}

	url := fmt.Sprintf("%s%s", t.domain, t.path.Pod)
	url = strings.Replace(url, "{external_id}", orderId, 1)

	resp, err := t.client.Get(url, true, t.audience)
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		errorResponse = responses.NewError(fmt.Sprintf("Error while reading response body: %s", err), "", "", t.logger)
		return
	}

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		errorResponse = responses.NewError(fmt.Sprintf("Received error: %s", body), strconv.Itoa(resp.StatusCode), "", t.logger)
		return
	}

	err = json.Unmarshal(body, &successResponse)
	if err != nil {
		errorResponse = responses.NewError(fmt.Sprintf("Error while unmarshalling response body: %s", err), "", "", t.logger)
		return
	}

	return
}
